GENOME_FA = "/diskmnt/Datasets/Reference/human_genome/GRCh37-lite/GRCh37-lite.fa"
BAM = "/diskmnt/Datasets/TCGA/example_BAM/TCGA-73-4677-11A-01D-1203-02_IlluminaHiSeq-DNASeq_whole.bam"
INTERVALPREFIX = "/diskmnt/Projects/Users/wliang/Germline_Noncoding/06_Cloud_Variant_Calling/RegulatoryGermline/reference/interval_chr"
chr=[str(i) for i in range (1,22)]
chr.append("X")
chr.append("Y")
print(chr)
rule gatk_haplotypecaller:
    input: 
        genome_fa = GENOME_FA,
        bam = BAM,
        interval = expand(f"{INTERVALPREFIX}{{ix}}.list", ix=chr),
    output: expand(f"gatk.raw.chr{{ix}}.vcf", ix=chr)
    shell: "gatk HaplotypeCaller -R {input.genome_fa} -I {input.bam} -L {input.interval} -O {output} --standard-min-confidence-threshold-for-calling 30.0"

rule gatk_selectvariant_snp:
    input:
        genome_fa = GENOME_FA,
        input_vcf = expand(f"gatk.raw.chr{{ix}}.vcf", ix=chr)
    output: expand(f"gatk.snp.chr{{ix}}.vcf", ix=chr)
    shell: "gatk SelectVariants -R {input.genome_fa} -V {input.input_vcf} -O {output} -select-type SNP -select-type MNP"

rule gatk_selectvariant_indel:
    input:
        genome_fa = GENOME_FA,
        input_vcf = "gatk.raw.vcf"
    output: "gatk.indel.vcf"
    shell: "gatk SelectVariants -R {input.genome_fa} -V {input.input_vcf} -O {output} -select-type INDEL"
